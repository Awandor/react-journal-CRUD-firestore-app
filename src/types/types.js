export const types = {
  login: '[Auth reducer] Login',
  logout: '[Auth reducer] Logout',

  uiSetError: '[UI reducer] set error',
  uiRemoveError: '[UI reducer] remove error',

  uiStartLoading: '[UI reducer] Start loading',
  uiFinishLoading: '[UI reducer] Finish loading',

  notesAddNewEntry: '[notes reducer] Add new entry',
  notesSetActive: '[notes reducer] Set active note',
  notesLoadEntries: '[notes reducer] Load all notes',
  notesUpdated: '[notes reducer] Updated notes',
  notesFileUrl: '[notes reducer] Updated image url',
  notesDeleteEntry: '[notes reducer] Delete entry',
  notesLogoutCleaning: '[notes reducer] Logout Cleaning'
};
