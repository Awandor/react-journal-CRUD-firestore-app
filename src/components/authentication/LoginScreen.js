import React from 'react';
import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { useDispatch, useSelector } from 'react-redux';
import { /* loginAction,  */startGoogleLoginAction, startLoginEmailPasswordAction } from '../../actions/authActions';

export const LoginScreen = () => {

    const [ formValues, handleInputChange ] = useForm( {
        email: 'test@test.es',
        password: '123456'
    } );

    const { email, password } = formValues;

    // Vamos a leer una propiedad del state

    const { loading } = useSelector( state => state.ui );

    const dispatch = useDispatch();

    const handleLogin = ( e ) => {

        e.preventDefault();

        // console.log( email, password );

        // Vamos a hacer el dispatch al Store

        // dispatch( loginAction( `rtge453333`, 'Dan' ) );

        dispatch( startLoginEmailPasswordAction( email, password ) );

    };

    const handleGoogleLogin = () => {

        console.log( 'handleGoogleLogin' );

        dispatch( startGoogleLoginAction() );

    };

    return (
        <>
            <h3 className="auth__title mb-5">Login Screen</h3>

            <form onSubmit={handleLogin}>
                <input className="auth__input" type="text" name="email" value={email} placeholder="Email" autoComplete="off" onChange={handleInputChange} />
                <input className="auth__input" type="password" name="password" value={password} placeholder="Password" onChange={handleInputChange} />

                <button className="btn btn-primary btn-block mb-5" type="submit" disabled={ loading }>Login</button>

                <hr />

                <div className="auth__social-networks">
                    <p>Login with social network</p>

                    <div className="google-btn" onClick={handleGoogleLogin}>
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <Link to="/auth/register" className="link">Create new account</Link>
            </form>
        </>
    );

};
