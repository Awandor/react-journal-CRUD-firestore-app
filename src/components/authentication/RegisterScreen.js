import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { startRegisterEmailPasswordNameAction } from '../../actions/authActions';
import { removeErrorAction, setErrorAction } from '../../actions/uiActions';
import { useForm } from '../../hooks/useForm';

export const RegisterScreen = () => {

    const [ formValues, handleInputChange ] = useForm( {
        name: 'Dan Anders Häggblom',
        email: 'test@test.es',
        password: '123456',
        password2: '123456'
    } );

    const { name, email, password, password2 } = formValues;

    // Podríamos usar useState para algo tan sencillo que sólo va usarse en esta página, pero usaremos Redux para aprender

    const dispatch = useDispatch();

    // Vamos a recuperar el mensaje de error del State

    /* const state = useSelector( state => state.ui );

    console.log( state ); */

    const { msgError } = useSelector( state => state.ui );

    console.log( 'msgError', msgError );

    const handleRegister = ( e )=>{

        e.preventDefault();

        // console.log( name, email, password, password2 );

        if( isFormValid() ){

            console.log( 'Formulario correcto' );

            dispatch( startRegisterEmailPasswordNameAction( email, password, name ) );
        
        }
    
    };

    const isFormValid = () => {

        if( name.trim().length === 0 ){

            console.log( 'Name is required' );

            dispatch( setErrorAction( 'Name is required' ) );

            return false;
        
        } else if( !validator.isEmail( email ) ){

            console.log( 'Email is not valid' );

            dispatch( setErrorAction( 'Email is not valid' ) );

            return false;
        
        } else if( password !== password2 || password.length < 5 ){

            console.log( 'Passwords are not equal or length must be 5 or longer' );

            dispatch( setErrorAction( 'Passwords are not equal or length must be 5 or longer' ) );

            return false;
        
        }

        // Remove error

        dispatch( removeErrorAction() );

        return true;

    };

    return (
        <>
            <h3 className="auth__title mb-5">Register Screen</h3>

            <form onSubmit={handleRegister}>

                {msgError && ( <div className="auth__alert-error">
                    {msgError}
                </div> )}

                <input className="auth__input" type="text" name="name" value={name} placeholder="Name" autoComplete="off" onChange={handleInputChange} />
                <input className="auth__input" type="text" name="email" value={email} placeholder="Email" autoComplete="off" onChange={handleInputChange} />
                <input className="auth__input" type="password" name="password" value={password} placeholder="Password" onChange={handleInputChange} />
                <input className="auth__input" type="password" name="password2" value={password2} placeholder="Confirm password" onChange={handleInputChange} />

                <button className="btn btn-primary btn-block mb-5" type="submit">Register</button>

                <Link to="/auth/login" className="link">Already registered?</Link>
            </form>
        </>
    );

};
