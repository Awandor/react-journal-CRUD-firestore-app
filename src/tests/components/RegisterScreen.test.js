import { mount } from 'enzyme';
import { RegisterScreen } from '../../components/authentication/RegisterScreen';
import { Provider } from 'react-redux';

// Como el componente hace uso del State necesitamos trabajar con el mock

import configureStore from 'redux-mock-store'; //ES6 modules
import thunk from 'redux-thunk';
import { MemoryRouter } from 'react-router-dom';
import { types } from '../../types/types';

const middlewares = [ thunk ];
const mockStore = configureStore( middlewares );

// Creamos el mock Store

const initialState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
};

let store = mockStore( initialState );

// const wrapper = mount( <Provider store={store}><MemoryRouter><LoginScreen /></MemoryRouter></Provider> );
const wrapper = mount( <Provider store={store}><MemoryRouter><RegisterScreen /></MemoryRouter></Provider> );


describe( 'Pruebas en <RegisterScreen />', () => {

    test( 'el componente debería mostrarse correctamente', () => {

        expect( wrapper ).toMatchSnapshot();

    } );

    test( 'debe de hacer el dispatch de la acción respectiva', () => {

        const emailField = wrapper.find( 'input[name="email"]' ); // selección por name

        emailField.simulate( 'change', {
            target: {
                value: '',
                name: 'email'
            }
        } );

        // Simulamos el submit con simulate

        wrapper.find( 'form' ).simulate( 'submit', {
            preventDefault(){}
        } );

        const actions = store.getActions();

        // console.log( actions );

        expect( actions[0] ).toEqual( { type: types.uiSetError, payload: 'Email is not valid' } );
        
    } );

    test( 'debería mostrar la caja de alerta con el error', () => {

        // El state no cambia por simular submit porque el test no está conectado a los reducers
        // Por ello cambiamos manualmente el state
        
        const initialState = {
            auth: {},
            ui: {
                loading: false,
                msgError: 'Email is not valid'
            }
        };
        
        const store = mockStore( initialState );
        
        const wrapper = mount( <Provider store={store}><MemoryRouter><RegisterScreen /></MemoryRouter></Provider> );

        const errorBox = wrapper.find( '.auth__alert-error' );

        // console.log( errorBox );

        expect( errorBox.exists() ).toBe( true );

        expect( errorBox.text().trim() ).toBe( initialState.ui.msgError );
        
    } );
    
    

} );
