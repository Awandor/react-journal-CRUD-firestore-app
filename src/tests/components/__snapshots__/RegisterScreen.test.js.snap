// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <RegisterScreen /> el componente debería mostrarse correctamente 1`] = `
Array [
  <h3
    className="auth__title mb-5"
  >
    Register Screen
  </h3>,
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Dan Anders Häggblom"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="test@test.es"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="123456"
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>,
]
`;
